Parsers/ASimDns/Parsers/ASimDnsInfobloxNIOS.yaml

```diff
@@ -26,49 +26,47 @@ ParserQuery: |
-        DnsQueryTypeName:string " "
-        "response: "
-        rest_of_string:string
-    | extend
-        // -- multi-line mode (?m) needed in regex since samples included new lines. Might not exist in real logs
-        extract = extract_all(@"^(\w+) ([\w\+]+?)(?: |$)(?m)(.*?$)?", rest_of_string)[0]
-    | extend 
-        DnsResponseCodeName = tostring(extract[0]),
-        DnsFlags = tostring(extract[1]),
-        DnsResponseName = tostring(extract[2])
-    | extend 
-        EventSubType = "response"
-    | project-away extract, rest_of_string,SyslogMessage, ProcessName, ProcessID, Facility, SeverityLevel, HostName
+    | parse SyslogMessage with *
+        "client " SrcIpAddr: string
+        "#" SrcPortNumber: string
+        " " NetworkProtocol: string
+        ": query: " DnsQuery: string
+        " " DnsQueryClassName: string
+        " " DnsQueryTypeName: string
+        " response: " DnsResponseCodeName: string
+        " " DnsFlags: string
+    | extend DNSResourceRecordIndex= indexof(DnsFlags, " ")
+    | extend DNSResourceRecord =iif(DNSResourceRecordIndex != "-1", substring(DnsFlags, DNSResourceRecordIndex), "")
+    | extend DnsFlags =iif(DNSResourceRecordIndex != "-1", substring(DnsFlags, 0, DNSResourceRecordIndex), DnsFlags)
+    | extend EventSubType = "response"
+    | project-away DNSResourceRecordIndex,SyslogMessage, ProcessName, ProcessID, Facility, SeverityLevel, HostName
   };
   let request = (disabled: boolean=false) {
     Syslog 
     | where not(disabled)
     | where ProcessName == "named" and SyslogMessage has_all ("client", "query:") and SyslogMessage !has "response:"
-    // -- trim_start needed to remove optional @xxxx identifier.
-    | extend SyslogMessage = trim_start (@'^(?:[\w\: ]+)?client (?:@.*? )?', SyslogMessage)
-    // -- trim_end needed to remove optional repetative IP address at the end.
-    | extend SyslogMessage = trim_end (@' \([\d\.]+\)', SyslogMessage)
-    | extend SyslogMessage = replace_string (SyslogMessage, '\\ ', '@@@')
-    | parse SyslogMessage with
-        SrcIpAddr:string "#" 
-        SrcPortNumber:int *
-        "query: "
-        DnsQuery:string " "
-        DnsQueryClassName:string " "
-        DnsQueryTypeName:string " "
-        DnsFlags:string
+    | parse SyslogMessage with *
+        "client " restData: string
+    | extend restData = iif(restData startswith "@", (substring(restData, indexof(restData, " "))), restData)
+    | extend restData = replace_string(restData,"\\ ","@@@")
+    | parse restData with 
+        SrcIpAddr: string
+        "#" SrcPortNumber: string
+        " " *
+        "query: " DnsQuery: string
+        " " DnsQueryClassName: string
+        " " DnsQueryTypeName: string
+        " " DnsFlags: string
     | extend DnsQuery = replace_string (DnsQuery, '@@@', ' ')
+    | extend ServerIPAddressIndex= indexof(DnsFlags, " ")
+    | extend ServerIPAddress = iif(ServerIPAddressIndex != "-1", substring(DnsFlags, ServerIPAddressIndex), "")
```

Comment:
Few comments on the splitting of DnsFlags:
- The substring expressions keep an unneeded space at the beginning of the second field (as noted above).
- While this can be solved with fixing the indexes, there are two preferable options here:
  - Since you anyway use regex, extract_all is probably more efficient and easier to understand.
  - Use split to get the values, with the current replace_regex.

Reply:
Updated logic to remove extra space: iif(DvcIpAddrIndex != "-1", substring(DnsFlags, DvcIpAddrIndex+1), "")
I was using regex to only remove the parenthesis around 'DvcIpAddr' in some logs using replace_regex(ServerIPAddress,@"[()]",""). I changed it and now using trim_start and trim_end to remove the parenthesis. Do you think its better now or we should be using the original extract_all logic?

    | extend DvcIpAddr = trim_start("[(]",DvcIpAddr)
    | extend DvcIpAddr = trim_end("[)]",DvcIpAddr)

Also, I now noticed 'HostIP' value is same as above calculated 'DvcIpAddr' value. So, I guess we don't really need above logic right? We already have below logic in the parser:

DvcIpAddr=iff (HostIP_s == "Unknown IP", "", HostIP_s)

Reply:
1. trim functions are regex based, so a single extract_all will be more efficient and, I think, comprehendible. 
2. I think this is the reason I did not parse DvcIpAddr from the message int he first place :-). Yup. Not needed.

Reply:
Thanks Ofer. Removed logic where I was calculating 'DvcIpAddr'. Below line should be good for storing DnsFlags and dumping the later texts:

| extend DnsFlags= tostring((split(DnsFlags," "))[0])

for handling both scenarios:
1. client @0x7f1a891bbbb 192.46.117.8#28492 (xxxx.uksouth.cloudapp.azure.com): view Recursion View: query: xxxx.uksouth.cloudapp.azure.com IN A +E(0)DV (192.46.116.155)
2. query: info: client 1.2.3.4#7619: query: 10.10.10.10.in-addr.arpa IN PTR -E